pkg/storage/gcs.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/spf13/pflag"
+	"google.golang.org/api/option"
+)
+
+const (
+	gcsEndpointOption     = "gcs.endpoint"
+	gcsStorageClassOption = "gcs.storage_class"
+	gcsPredefinedACL      = "gcs.predefined_acl"
+	gcsCredentialsFile    = "gcs.credentials_file"
+)
+
+// GCSBackendOptions are options for configuration the GCS storage.
+type GCSBackendOptions struct {
+	Endpoint        string `json:"endpoint" toml:"endpoint"`
+	StorageClass    string `json:"storage_class" toml:"storage_class"`
+	PredefinedACL   string `json:"predefined_acl" toml:"predefined_acl"`
+	CredentialsFile string `json:"credentials_file" toml:"credentials_file"`
+}
+
+func (options *GCSBackendOptions) apply(gcs *backup.GCS) error {
+	if options.CredentialsFile == "" {
+		return errors.New("must provide 'gcs.credentials_file'")
+	}
+
+	gcs.Endpoint = options.Endpoint
+	gcs.StorageClass = options.StorageClass
+	gcs.PredefinedAcl = options.PredefinedACL
+
+	b, err := ioutil.ReadFile(options.CredentialsFile)
+	if err != nil {
+		return err
+	}
+	gcs.CredentialsBlob = string(b)
+	return nil
+}
+
+func defineGCSFlags(flags *pflag.FlagSet) {
+	flags.String(gcsEndpointOption, "", "Set the GCS endpoint URL")
+	flags.String(gcsStorageClassOption, "",
+		"Specify the GCS storage class for objects. "+
+			"If it is not set, objects uploaded are followed by the default storage class of the bucket. "+
+			"See https://cloud.google.com/storage/docs/storage-classes for valid values.")
+	flags.String(gcsPredefinedACL, "",
+		"Specify the GCS predefined acl for objects. "+
+			"If it is not set, objects uploaded are followed by the acl of bucket scope. "+
+			"See https://cloud.google.com/storage/docs/access-control/lists#predefined-acl for valid values.")
+	flags.String(gcsCredentialsFile, "",
+		"Set the GCS credentials file path. You can get one from https://console.cloud.google.com/apis/credentials.")
+
+	_ = flags.MarkHidden(gcsEndpointOption)
+}
+
+func getBackendOptionsFromGCSFlags(flags *pflag.FlagSet) (options GCSBackendOptions, err error) {
+	options.Endpoint, err = flags.GetString(s3EndpointOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	options.StorageClass, err = flags.GetString(gcsStorageClassOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	options.PredefinedACL, err = flags.GetString(gcsPredefinedACL)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+
+	options.CredentialsFile, err = flags.GetString(gcsCredentialsFile)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	return
+}
+
+type gcsStorage struct {
+	gcs    *backup.GCS
+	bucket *storage.BucketHandle
+}
+
+// Write file to storage
+func (s *gcsStorage) Write(ctx context.Context, name string, data []byte) error {
+	object := s.gcs.Prefix + name
+	wc := s.bucket.Object(object).NewWriter(ctx)
+	wc.StorageClass = s.gcs.StorageClass
+	wc.PredefinedACL = s.gcs.PredefinedAcl
+	_, err := wc.Write(data)
+	if err != nil {
+		return err
```

Comment:
Forget `wc.Close()`?

Reply:
It seems need not to Close here, see https://github.com/GoogleCloudPlatform/golang-samples/blob/master/storage/objects/main.go#L92-L110 as example.